benchmark/Rust/Savina/src/concurrency/Dictionary.lf

```diff
@@ -0,0 +1,167 @@
+ * The dictionary processes the requests in a fixed order which makes the whole
+ * application deterministic.
+ * 
+ * To break the causality loop the dictionary reactor contains a logical action.
+ * This appears to be more efficient than using a logical action within each
+ * worker. In a quick test, the version with logical actions in each worker was
+ * 50% slower compared to the version with only one logical action in the
+ * dictionary.
+ * 
+ * @author Christian Menard
+ * @author Hannes Klein
+ */
+
+target Rust {
+    build-type: Release,
+    rust-include: [ "dictionary_support.rs" ],
+    cargo-features: [ "cli" ],
+    cargo-dependencies: {
+        rand: {
+            version: "0.8",
+            features: ["small_rng"],
+        },
+        reactor_rt: {
+            features: ["parallel-runtime"]
+        },
+    }
+};
+
+import BenchmarkRunner from "../BenchmarkRunner.lf";
+
+reactor DictionaryImpl(num_workers: usize(20)) {
+
+    preamble {=
+        use crate::dictionary_support::*;
+        use std::collections::HashMap;
+        use std::convert::TryInto;
+    =}
+
+    state num_workers(num_workers);
+    state data_map: HashMap<u32, u32>;
+    state answers_to_send: Vec<i32> ({= vec![ -1 ; num_workers ] =});
+
+    state num_requests_served: u32(0);
+
+    // Having the action in the dictionary is faster... (this is a c++ comment)
+    logical action send_answers: {= () =};
+
+    input[num_workers] request: Message;
+    output[num_workers] response: u32;
+
+    // @label send_answers
+    reaction(send_answers) -> response {=
+        for i in 0..self.num_workers {
+             if let Ok(answer) = self.answers_to_send[i].try_into() {
+                ctx.set(response.get(i), answer);
+            }
+        }
+    =}
+    
+    // @label recv_request
+    reaction(request) -> send_answers {=
+        // The order of messages to read is relevant, it effectively
+        // assigns priorities to the workers.
+        for (i, request_i) in request.into_iter().enumerate() {
+            ctx.use_ref_opt(&request_i, |msg| {
+                self.num_requests_served += 1;
+                match msg {
+                    Message::Write { key, value } => {
+                        self.data_map.insert(*key, *value);
+                        self.answers_to_send[i] = (*value).try_into().unwrap_or(0);
+                        info!("Worker {} writes {} to key {}", i, value, key);
+                    },
+                    Message::Read { key } => {
+                        // read the value. If the key is not yet present, this will add it and initialize the value to 0.
+                        // This does not match 100% the Savina implementation which returns null if the key is not present,
+                        // but should do a similar job.
+
+                        // todo wat? does the savina impl insert a new (key->null) mapping or just return null?
```

Comment:
Reading a value that is not in the map, should just return null and don't insert a value. This was actually a bug in the C++ implementation.

Reply:
Ok, well the rust implementation does insert a mapping if one is missing. I added this behavior because otherwise we're always asking for random numbers which aren't in the map. Maybe it's because we use SmallRng and not the custom prng. I would assume then, that this implementation of Dictionary is not comparable to the c++ or Akka one. Probably it's the same thing with all benchmarks that use a prng (RadixSort and CigaretteSmokers). Should I still include them in my report?

Reply:
this would also explain why the curves for dictionary are different in the screenshots above